### ISSUE TITLE
Have jigsaw calculate the sparsity 

### ISSUE THREAD

**Issue by blandoplanet** [DOI-USGS/ISIS3#5177]
**Description**  
We have had multiple cases of large control networks fail in jigsaw while reporting a "problem too large" issue with cholmod (see [5173](https://github.com/DOI-USGS/ISIS3/issues/5173) and [3871](https://github.com/DOI-USGS/ISIS3/issues/3871). This problem occurs despite providing sufficient memory to solve the network. We have also found that some large networks solve without issue. For example, the new Kaguya TC southern hemisphere network with 81K images, 5.4M points, and 21M measures fails, but the CTX global with 100k images, 9.4M points, and 30M measures solves. Before he departed USGS, Jesse suggested that the size of the network is less important than the sparsity of the resulting matrix. The more connections, the less sparse the matrix. Our former photogrammetrist has suggested the same. This is consistent with the observation that all of the failing networks contain the pole, which has a lot of well-connected data (high sparsity?), but large networks that solve successfully (e.g., CTX) do not contain the pole (low sparsity?).  We find, for example, that removing the polar quad from the Kaguya TC southern hemisphere network results in jigsaw solving successfully, despite still being "large" (66K images, 4.8M points, 12M measures).

The sparsity of a matrix is simply the fraction of zero elements in the matrix. I.e., the number of zero elements / the total number of elements. Having jigsaw calculate and report the sparsity of the matrix before inversion would help us diagnose why some networks are failing, and could allow us to develop rules of thumb regarding when the sparsity of a matrix is too low.

**Example**  
Conceptually this seems easy to do. In reality it may not be - I do not have any deep knowledge on this topic.

**Comment by blandoplanet** [comment]
This work would still be valuable.

**Comment by lwellerastro** [comment]
I have two small test cases available under my user work area Isis3Tests/Jigsaw/Sparsity/:
Galileo/ - 4 images, all points are 2 measures deep, overlaps near image corners
KaguytaTC - 21 images, max point depth is 12, lots of overlaps and connectivity (I think)

I would think Galileo is more "sparse" than the Kaguya network, but I don't really know.  

Both have images which are local under the Lev0/ directory, they have a jigsaw command in the file proc.scr and output files for reference.  The Kaguya data set and network is the same used for an initial small cholmod 64 test and you might already have it.

**Comment by AustinSanders** [comment]
I looked into this, and it looks like sparsity calculation isn't quite as easy as n_zeros/matrix_size in practice, because we're never generating a "full" matrix.  Jigsaw uses piecewise calculations, sparse block matrices, and cholmod triplets to perform the calculation, and I'm having trouble determining the size of the "full" matrix because it never exists all at once.  It does look like the total number of nonzero elements is exposed, though.

This got me thinking:

Is it really the sparsity that dictates the success / failure of the bundle adjustment?  Sparsity certainly plays a role, but only insofar as that it relates to the total number of connections.  It would seem that a tiny network would be allowed more sparsity than a giant network, i.e. a trivially small network could have a sparsity of 0 and still succeed, so the sparsity alone would seem to be an insufficient metric for determining what will / will not work in jigsaw.  It seems to me that the driving force behind the success / failure of a network is more closely related to the total number of nonzero elements, not the relative proportion of zero/nonzero elements.

Would reporting the total number of non-zero elements provide enough information to accomplish the goal here, or is there a strong reason to convert it into a sparsity value?

---

TLDR:
Sparsity is kind of hard to calculate because of the specific data structure / process we're using, and I'm not sure that it's the most important metric anyway.  What if we reported the total number of nonzero elements instead?

**Comment by lwellerastro** [comment]
I honestly don't know @AustinSanders - so out of my league. But maybe @kledmundson would be willing to comment?

**Comment by kledmundson** [comment]
Will respond after I've reviewed my notes.

**Comment by lwellerastro** [comment]
sure thing @kledmundson, if you have time to think about it. Mostly was wondering about Austin's question about reporting total number of nonzero elements instead of sparsity. Any additional information in regard to a networks complexity seems useful, but what metric we use is not something I can help determine.

**Comment by kledmundson** [comment]
1) the number of elements in the full sized reduced normals is m_rank * m_rank.
(m_rank is a member of BundleAdjust)

2) It _is_ more straightforward to get the number of nonzero elements since we've smushed out all of the zeros, so...

3) the "density" number (for lack of a better term) would be 

    nonzero elements / (m_rank * m_rank)


4) **But**, if you've got the "density" number, then you've got the sparsity number by default cause it's just 1 - the "density" number


5) I would lean toward reporting the sparsity number (or possibly also expressed as a percentage) if for no other reason because of convention


6) What might seem to be tricky here is getting the number of nonzero elements in the full matrix cause my recollection is that the cholmod triplet is storing the upper (or maybe lower) triangular portion of the matrix. But I suspect that's trivial as well. Doing a quick check on that and will be back shortly (I hope) with an answer.

**Comment by AustinSanders** [comment]
That's good information -- I think the insight on m_rank*m_rank will make a big difference.

That being said, before more work goes into calculating the sparsity using the upper / lower triangle, I still have concerns regarding the suitability of using sparsity for what we're trying to accomplish.  

A 3 image network with sparsity .5 and a 3,000,000 image network with sparsity .5 are very different.  One will succeed and one will fail, but the same sparsity is reported for each network, so it doesn't seem like the right metric with which to "diagnose why some networks are failing and allow us to develop rules of thumb...."

It seems to me that using sparsity in conjunction with network size is simply a roundabout way of gauging total number of nonzero elements, which is what ultimately impacts memory consumption.

**Comment by kledmundson** [comment]
While I think reporting sparsity is generally of value, I don't disagree with what you're saying regarding the original point of the issue.

If there is doubt about whether sufficient memory is available for the solution, it may be valuable to at least approximate the memory required and compare it to what's available before solving.

But that's not just a function of the size of the reduced normal equations matrix because there are auxiliary matrices generated for each point in addition to the normals matrix. And that may not be insignificant for networks with a lot of points.

Maybe this one becomes less of an immediate priority since the S. Pole net problem wasn't ultimately due to insufficient memory?

**Comment by AustinSanders** [comment]
The dev team has discussed this issue, and the consensus is that sparsity is no the correct metric to diagnose network failure.

We've talked about closing this issue, but I think the heart of the issue is less about "we need a sparsity calculation," and more of "we need some reportable metric to help understand why networks are failing."

@blandoplanet Given that the S. Pole network problem wasn't a memory issue, is working on finding + reporting a meaningful metric still valuable?

**Comment by blandoplanet** [comment]
@AustinSanders I agree. The real motivation is/was indeed "we need some reportable metric to help understand why networks are failing." At the time, it appeared that polar networks might be failing due to their "complexity" relative to equatorial networks, and sparsity seemed to be a way to evaluate that. I honestly have not done my homework to know what the right metric(s) are and it may be that the sparsity is not the right path. It seemed "easy" to calculate so appeared to be low hanging fruit, but this is not the case.

Given that jigsaw itself is now running fine, I think it would be more valuable to work on other tools that can help us better understand network quality (e.g., determining how well the network is connected).
